src/runtimes/node.js

```diff
@@ -116,30 +117,47 @@ const zipFunction = async function ({
     ignoredModules: [...ignoredModulesFromConfig, ...ignoredModulesFromSpecialCases],
     srcFile: mainFile,
   })
+  const bundlerWarnings = data.warnings.length === 0 ? undefined : [data.warnings]
 
   // We're adding the bundled file to the zip, but we want it to have the same
-  // name and path as the original, unbundled file. For this, we use a rename.
-  const renames = {
-    [bundlePath]: mainFile,
+  // name and path as the original, unbundled file. For this, we use an alias..
+  const aliases = {
+    [mainFile]: bundlePath,
   }
   const basePath = commonPathPrefix([...dirnames, dirname(mainFile)])
 
   try {
     await zipNodeJs({
+      aliases,
       basePath,
       destFolder,
       destPath,
       filename,
       mainFile,
       pluginsModulesPath,
-      renames,
-      srcFiles: [...srcFiles, bundlePath],
+      srcFiles,
     })
   } finally {
     await cleanTempFiles()
   }
 
-  return destPath
+  return { bundlerVersion: JS_BUNDLER_ESBUILD, bundlerWarnings, path: destPath }
+}
+
+const zipWithFunctionWithFallback = async (parameters) => {
+  // If a specific JS bundler version is specified, we'll use it.
+  if (parameters.jsBundler) {
+    return zipFunction(parameters)
+  }
+
+  // Otherwise, we'll try to bundle with v2 and, if that fails, fallback to v1.
+  try {
+    return await zipFunction({ ...parameters, jsBundler: JS_BUNDLER_ESBUILD })
+  } catch (error) {
+    const data = await zipFunction({ ...parameters, jsBundler: JS_BUNDLER_LEGACY })
+
+    return { ...data, bundlerErrors: error.errors }
```

Comment:
Are there some `esbuild` exceptions which do not include the `errors` property? If so, do we want to capture those too?

Generally speaking, should we propagate any exception from `zipFunction()`? In other words, returning `error` instead of `error.errors`. Otherwise, we would not get any error message nor stack trace form bugs in `zip-it-and-ship-it`.

Reply:
This ties back to the fact that `zip-it-and-ship-it` still technically supports a binary entry point with a JSON output, so any non-error states should be serialisable to JSON. In this case, we're returning errors from esbuild but the status of the application is not an error state — i.e. esbuild has failed, but we were able to succeed in zipping the functions thanks to the fallback.

esbuild returns an `errors` and `warnings` properties, containing information such as the location (file, line and column) and the description of the event, which I was thinking of surfacing in Netlify Build.

Does that make sense?

Reply:
Yes, thanks for explaining :+1: 

Does `error.stack` contains that information as well? If so, it could also be used, and is JSON-serializable.